op-node-derivers.md

```diff
@@ -0,0 +1,554 @@
+
+By categorizing events, we can schedule processing in parallel, without enshrining specific resources:
+```go
+type Resource string
+
+const (
+    UnsafeL2 Resource = "unsafe-l2"
+    PromoteUnsafeL2 Resource = "promote-unsafe-l2" // example: to not tightly couple "unsafe" to "safe" transition
+    SafeL2 Resource = "safe-l2"
+    FinalizedL2 Resource = "finalized-l2"
+    CurrentL1 Resource = "current-l1" // traversal point on L1 for block-attributes generation
+  
+    // ... standard resources can be extended by mods / new projects, e.g.:
+    CrossUnsafeL2 Resource = "cross-unsafe-l2"
+    
+    // some functionality may gets its own category, to synchronize the work into one place, for less lock contention
+    EngineL2 Resource = "engine-l2"
+)
+
+type Event struct {
+    Resource Resource
+    Data any
+}
+```
+
+We can introduce a `Deriver` as an event processor:
+```go
+type Deriver interface {
+    OnEvent(ctx context.Context, event *Event)
+}
+```
+A deriver is assumed to read-access to any state it needs. Composed by the constructor of the deriver.
+
+#### Composable derivers
+
+Writes idiomatically happen by implementing a purpose-specific Deriver:
+```go
+type Engine struct {
+  control EngineController
+}
+
+func (e *Engine) OnEvent(ctx context.Context, event *Event) {
+    switch x := event.Data.(type) {
+        case *NewPayload:
+            select {
+              // events can include ways to communicate the result, if necessary, across threads
+              case x.Result <- e.onNewPayload(x.Envelope):
+              case <-ctx.Done():
+            }
+        // *insert other engine events*
+    }
+}
+
+func (e *Engine) onNewPayload(env *eth.ExecutionPayloadEnvelope) error {
+    // ... check prestate first etc., insofar the EngineController does not already maintain it.
+    return e.control.InsertUnsafePayload(ctx, x.PayloadEnvelope)
+}
+```
+
+Other derivers can then re-use this, by composing:
+```go
+type Rollup struct {
+    engine Deriver
+    pipeline Deriver
+    // ...
+}
+
+func (r *Rollup) OnEvent(ctx context.Context, event *Event) {
+  switch x := event.Data.(type) {
+  case *NewPayload:
+      // 1) update some rollup state: ...
+      // 2) and propagate to the engine deriver
+      r.engine.OnEvent(ctx, event)
+  case *L1Change:
+      r.pipeline.OnEvent(ctx, event)
+  default:
+      r.pipeline.OnEvent(ctx, event) // can forward anything to a fallback
+  }
+}
+```
+
+Or, a deriver can be given access to the root-level Deriver, to surface new event data to:
+```go
+type Pipeline struct {
+    root Deriver
+}
+
+func (p *Pipeline) OnEvent(ctx context.Context, event *Event) {
+    switch x := event.Data.(type) {
+    case *L1Change:
+        // generate block attributes and give them to the unsafe-block processor to extend the chain
+        // (or forward to next deriver if already known attributes)
+        p.root.OnEvent(ctx, &Event{Resource: UnsafeL2, Data: AttributesEvent{attributes}})
```

Comment:
It would definitely be very useful to have a clear set of threading expectations around event handling.  For example does `OnEvent` have to be handled synchronously or could it be async? Is that deriver dependent?

If async is allowed and/or expected, then a deriver that isn't thread safe could just be wrapped in a deriver that executes all events on a single thread (either by channels or with a lock).

Reply:
Tried to answer in aggregate resource/event comment below. Let's discuss there, this discussion is getting a bit sparse but it's all related.